wgpu/src/backend/wgpu_core.rs

```diff
@@ -736,7 +741,19 @@ impl crate::Context for ContextWgpuCore {
                     },
                 )
             }
-            Err(err) => self.handle_error_fatal(err, "Surface::get_current_texture_view"),
+            Err(err) => match surface_data.error_sink.lock().as_ref() {
+                Some(error_sink) => {
+                    self.handle_error_nolabel(error_sink, err, "Surface::get_current_texture_view");
```

Comment:
thanks for sharing those callstacks. This kind of tracking is incredibly valuable. We do some of that at [Rerun](https://rerun.io/) as well but our userbase is a lot smaller at this point and a lot of the traffic is blocked (and it's easy to opt out), we don't have that good of an insight.

> The most ergonomic solution to this (IMO, obviously you know this code better than I do) would be to forward the error and return a result in both functions so the caller can handle them. As you mentioned in the issue when we were discussing earlier, we need to conform with the webgpu spec around handling errors asynchronously so I'm not sure that's totally feasible.

100% agree the more we can forward directly and still be WebGPU compatible the better!
Seeing the picture a bit better by now, I think I actually have to row back a bit on my statements about that we can't return Results at all:
* whenever webgpu spec says that something **throws** on the **content timeline** (the green marked stuff), we can catch those failures and convert them to a Result
     * the "queue up the error" case is for every failure on the device timeline and that is indeed most of the validation and why unfortunately a ton of things can't be Results and instead are asynchronously reported errors
* whenever a failure (or better, a family of errors) occurs that can only happen with wgpu-core and never with the js webgpu api, then we can return a result

[get_current_texture](https://www.w3.org/TR/webgpu/#dom-gpucanvascontext-getcurrenttexture) in particular has practically no failure case other than some resources being null which afaik isn't possible in our Rust interface. 

> What I'm find awkward about the current approach (and I think this is partly to your point) is that the get_current_texture already returns a Result

Yes agree again. With the above in mind about what failures are possible in wgpu-core vs webgpu this makes a bit more sense.

> I think what is in this PR is ok (and acceptable) 

Yep, definitely an improvement. Let's not get perfect in the way of good 😄 

> We can't return Ok (because we don't have a SurfaceTexture) but we could return an actual Err with the details of what happened (instead of just SurfaceError::Unknown). I'm not sure that's that much better though since we'd duplicate the error across the return type and the error callbacks...Is there prior art for error handling like this in wgpu that you know of? Would love suggestions on this :D

I think we should go over all surface relateed results & error codes again across the crate and unify them as much as possible. I haven't thought it through yet, but I think we could have status codes only show up as part of usable textures and everything else be an error with no status code at all.
Right now there's a lot of weird transformation going on between the wgpu interface, the wgpu-core context (still inside the wgpu crate) and the wgpu-core crate which has the actual implementation of the surfaces. Ideally we just pass things through as-is.

Not a lot of prior art to this unfortunately. There should probably more places in wgpu where we give back some limited Result - grepping for `throw` in the webgpu spec to see where the content timeline can fail yields a handful of results and they should probably all translate into `Result`. Very few do today, buffer mapping has some iirc.